src/UCTNode.h

```diff
@@ -24,20 +24,97 @@
 #include <atomic>
 #include <memory>
 #include <vector>
+#include <cassert>
+#include <cstring>
 
 #include "GameState.h"
 #include "Network.h"
 #include "SMP.h"
 
+class UCTNode;
+
+// 'lazy-initializable' version of std::unique_ptr<UCTNode>.
+// When a UCTNodePointer is constructed, the constructor arguments are stored instead
+// of constructing the actual UCTNode instance.  Later when the UCTNode is needed,
+// the external code calls expand() which actually constructs the UCTNode.
+// Basically, this is a 'tagged union' of:
+//  - std::unique_ptr<UCTNode> pointer;
+//  - std::pair<float, std::int16_t> args;
+
+// WARNING : expand() is not thread-safe and hence has to be protected by an external lock.
+
+class UCTNodePointer {
+private:
+    // the raw storage used here.
+    // if bit 0 is 0, m_data is the actual pointer.
+    // if bit 0 is 1, bit [31:16] is the vertex value, bit [63:32] is the score.
+    // (I really wanted to use C-style bit fields and unions, but those aren't portable)
+    mutable uint64_t m_data = 1;
```

Comment:
Is the reason for the use of bit-fields and unions not being portable because this "is expanded" flag's bit location is implementation-specific and/or are there other reasons?
Would storing the flag "is expanded" elsewhere (e.g. packed in a bit-field with `m_is_expanding`) allow your desired approach to be used here?

Reply:
Yeah, the reason was that bit positions are not portable.  I couldn't find a suitable library on standard C++ nor it seemed worth bringing in another external dependency for such a small amount of code.  Any suggestions to make it prettier are welcome :)

Reply:
Thanks for confirming my understanding. 
I can't tell from your reply whether you have considered:
> Would storing the flag "is expanded" elsewhere (e.g. packed in a bit-field with m_is_expanding) allow your desired approach to be used here?

Reply:
Elsewhere... means where?  Adding a separate field ("bool m_is_expanding") on the UCTNodePointer will add another byte of memory on the eight byte field, making it a 9-byte structure (which probably will turn into 12 or 16 on modern compilers).

I wasn't considering adding more memory footprint simply because if I wanted to, I could avoid using the union mess altogether and implement it as sizeof(UCTNodePointer) == 16 instead of 8 (and spending more memory).

Reply:
I should have made it clearer than I meant within the `UCTNode` class, next to the existing `m_is_expanding` that is currently taking up a whole byte for a single bit. Something like:
```
    ...
    // Is someone adding scores to this node?
    // We don't need to unset this.
    bool m_is_expanded : 1;
    bool m_is_expanding : 1;
    SMP::Mutex m_nodemutex;
    ...
```